### PR TITLE
New Feature: Enable/Disable Roles

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR allows enabling/disabling roles. By default all the existing roles are set as 'enabled'.

New APIs created:
- enableRole
- disableRole

The listRoles API is extended to support a new parameter: state.

Fixes: #8460 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):
Screenshot 2024-08-20 at 00 22 58

### How Has This Been Tested?
Create a role, verify role is enabled
Disable role
Add account, verify the disabled role is not listed

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by JoaoJandre** [comment]
Both APIs basically do the same thing, they just toggle the same attribute. We should have a single API for this feature, we could call it `setRoleState` for example.

**Comment by nvazquez** [comment]
@JoaoJandre thanks you are right, I even thought on including it as part of the updateRole API but kept it separate following the same pattern for accounts and users, but I'm fine with the unified approach as well

**Review by kiranchavala — APPROVED** [review]
LGTM

Before fix 
```

(cmk) > list roles -h
listRoles: Lists dynamic roles in CloudStack
API Params               Type     Description
==========               ====     ===========
id                       uuid     List role by role ID.
keyword                  string   List by keyword
name                     string   List role by role name.
page                     integer
pagesize                 integer
type                     string   List role by role type, valid options are
                                  : Admin, ResourceAdmin, DomainAdmin, Use
                                  r.
```



After fix 


```
(cmk)  > list roles -h
listRoles: Lists dynamic roles in CloudStack
API Params               Type     Description
==========               ====     ===========
id                       uuid     List role by role ID.
keyword                  string   List by keyword
name                     string   List role by role name.
page                     integer
pagesize                 integer
state                    string   List role by role type status, valid opti
                                  ons are: enabled, disabled
type                     string   List role by role type, valid options are
                                  : Admin, ResourceAdmin, DomainAdmin, Use
                                  r.

```

```
(cmk) 🐱 > list roles name=test
{
  "count": 1,
  "role": [
    {
      "id": "176ecc6f-6627-4369-b752-f6d9f2d6bc56",
      "isdefault": false,
      "ispublic": true,
      "name": "test",
      "state": "disabled",
      "type": "User"
    }
  ]
}
(cmk) 🐱 > list roles name=test
{
  "count": 1,
  "role": [
    {
      "id": "176ecc6f-6627-4369-b752-f6d9f2d6bc56",
      "isdefault": false,
      "ispublic": true,
      "name": "test",
      "state": "enabled",
      "type": "User"
    }
  ]
}

```

Screenshot 2024-08-21 at 3 22 41 PM

Screenshot 2024-08-21 at 3 31 16 PM

File: engine/schema/src/main/java/org/apache/cloudstack/acl/RoleVO.java [hunk 1]
```diff
@@ -58,11 +58,16 @@ public class RoleVO implements Role {
...
+    @Column(name = "state")
+    @Enumerated(value = EnumType.STRING)
+    private State state;
+
     @Column(name = GenericDao.REMOVED_COLUMN)
     private Date removed;
 
     public RoleVO() {
         this.uuid = UUID.randomUUID().toString();
+        this.state = State.ENABLED;
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply engine/schema/src/main/java/org/apache/cloudstack/acl/RoleVO.java:70]: well, I'd expect the DAO to instantiate a unmarshalled Role with state disabled if it was so when saved. But maybe accessors are enough for that.

**Comment by nvazquez** [comment]
@JoaoJandre can we address the unification of the enable/disable APIs for roles along with accounts and users in a separate PR? Possibly there could be other resources following the same pattern that we could investigate and unify

**Comment by JoaoJandre** [comment]
> @JoaoJandre can we address the unification of the enable/disable APIs for roles along with accounts and users in a separate PR? Possibly there could be other resources following the same pattern that we could investigate and unify

@nvazquez I'm ok with the idea. However, we still do not have a proper mechanism for deprecating/changing API's names (see https://github.com/apache/cloudstack/discussions/8970). The discussion has died out a bit, but I think it's still worth it to try to create a consensus on how to do this, vote it and establish as a practice on the project.

Going back to your idea, currently, as we do not have a clear way of changing API names or removing APIs, I would prefer if we introduced these APIs as one consolidated API from the beginning.

**Comment by DaanHoogland** [comment]
> > @JoaoJandre can we address the unification of the enable/disable APIs for roles along with accounts and users in a separate PR? Possibly there could be other resources following the same pattern that we could investigate and unify
> 
> @nvazquez I'm ok with the idea. However, we still do not have a proper mechanism for deprecating/changing API's names (see #8970). The discussion has died out a bit, but I think it's still worth it to try to create a consensus on how to do this, vote it and establish as a practice on the project.
> 
> Going back to your idea, currently, as we do not have a clear way of changing API names or removing APIs, I would prefer if we introduced these APIs as one consolidated API from the beginning.

@JoaoJandre I would prefer we do not introduce a new pattern for related entities unless we are sure we will address all of them. I think it is best to leave this PR as is, as it adheres better to related code/interfaces. I don't want to be strict in this perspect but @nvazquez ' argument makes sense to me.

**Pull request merged**